### PR TITLE
[consensus] sync improvements to help slow nodes sync better

### PR DESCRIPTION
**Pull request by ibalajiarun**
## Description

- In Quorum store, we garbage collect batches 60 seconds after they are expired based on block timestamp, to help slow peers fetch batches and execute without falling back to state sync.
- In consensus, peers will state sync if their local committed block timestamp is behind 30s from the remote committed block timestamp.

### CODE REVIEW

File: consensus/src/block_storage/sync_manager.rs [hunk 1]
```diff
@@ -68,6 +68,13 @@ impl BlockStore {
             && !self.block_exists(li.commit_info().id()))
             || self.commit_root().round() + 30.max(2 * self.vote_back_pressure_limit)
                 < li.commit_info().round()
+            // If the LI commit block timestamp is more than 60 secs ahead of self commit block 
+            // timestamp, sync to the ledger info
+            || li
+                .commit_info()
+                .timestamp_usecs()
+                .saturating_sub(self.commit_root().timestamp_usecs())
+                >= Duration::from_secs(60).as_micros()
```
Comment (zekun000) [inline consensus/src/block_storage/sync_manager.rs:77]: think more about it, I'd like to have this number smaller than the one below for more buffer as well, maybe 30s here?

File: consensus/src/quorum_store/batch_store.rs [hunk 2]
```diff
@@ -283,15 +283,20 @@ impl BatchStore {
...
+        let expired_digests = self.expirations.lock().unwrap().expire(expiration_time);
         let mut ret = Vec::new();
         for h in expired_digests {
             let removed_value = match self.db_cache.entry(h) {
                 Occupied(entry) => {
                     // We need to check up-to-date expiration again because receiving the same
                     // digest with a higher expiration would update the persisted value and
                     // effectively extend the expiration.
-                    if entry.get().expiration() <= certified_time {
+                    if entry.get().expiration() <= expiration_time {
```
Comment (zekun000) [inline consensus/src/quorum_store/batch_store.rs:299]: not sure how this affect the quotas and do we need to adjust any numbers

Comment (ibalajiarun) [reply consensus/src/quorum_store/batch_store.rs:299]: We will need 24mb per peer to store 120 seconds of data. I think this is acceptable.

**Review by zekun000 — APPROVED** [review]

File: consensus/src/block_storage/sync_manager.rs (hunk 1 above)
Comment (ibalajiarun) [reply consensus/src/block_storage/sync_manager.rs]: Hmm, there is still a 60s grace period to sync to higher LI and still have the payload available right?

File: consensus/src/quorum_store/batch_store.rs [hunk 3]
```diff
@@ -283,15 +286,19 @@ impl BatchStore {
     // pub(crate) for testing
     #[allow(clippy::unwrap_used)]
     pub(crate) fn clear_expired_payload(&self, certified_time: u64) -> Vec<HashValue> {
-        let expired_digests = self.expirations.lock().unwrap().expire(certified_time);
+        // To help slow nodes catch up via execution without going to state sync we keep the blocks for 60 extra seconds
+        // after the expiration time. This will help remote peers fetch batches that just expired but are within their
+        // execution window.
```
Comment (bchocho) [inline consensus/src/quorum_store/batch_store.rs:291]: "execution window" lol

**Review by msmouse — APPROVED** [review]
stamp

File: consensus/src/quorum_store/batch_store.rs [hunk 4]
```diff
@@ -283,15 +286,19 @@ impl BatchStore {
...
+        let expiration_time = certified_time.saturating_sub(self.expiration_buffer_usecs);
```
Comment (bchocho) [inline consensus/src/quorum_store/batch_store.rs:292]: would `adjusted_certified_time` be easier to understand? but maybe it's just me

**Review by bchocho — APPROVED** [review]
LGTM

**Pull request merged**